### ISSUE TITLE
SFTP publickey authentication fails unexpectedly when user has no shadow password info

### ISSUE THREAD

**Issue by jpfinger** [proftpd/proftpd#890]
### What I Did

After upgrading from 1.3.5d to 1.3.6c (git 7f2f0ab), SFTP accounts no longer work with public key authentication.

The environment has two authentication sources, a AuthUserFile and SSSD (FreeIPA).  Most users login by password with SSSD.  Some users login by password with the AuthUserFile.  Yet other users login by public keys (no password) using SSSD for group enumeration, etc.

What still works for SFTP after the update:

- Username+password in AuthUserFile only
- Username+password in SSSD only
- Username+password in AuthUserFile and SSSD; different passwords both work
- Public key where the user exists in AuthUserFile and SSSD
- Username+password in SSSD where an authorized_key exists but not used

What does not work after the update:

- Public key authentication where the user exists in SSSD only

In the attached sftp trace log we see it successfully authenticate but also fail:

```
2020-01-30 18:49:38,631 [15300] <ssh2:8>: verified public key for user 'PubKeyUser'
2020-01-30 18:49:38,631 [15300] <ssh2:19>: waiting for max of 600 secs while polling socket 1 for writing using select(2)
2020-01-30 18:49:38,631 [15300] <ssh2:3>: sent SSH_MSG_USER_AUTH_FAILURE (51) packet (80 bytes)
2020-01-30 18:49:38,631 [15300] <ssh2:19>: waiting for max of 600 secs while polling socket 0 for reading using select(2)
2020-01-30 18:49:38,632 [15300] <ssh2:16>: disconnecting client (received EOF)
```

OS: CentOS 7.7.1908 kernel 3.10.0-1062.9.1.el7.x86_64 #1 SMP Fri Dec 6 15:49:49 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

### What I Expected/Wanted

Users able to login with user/group information in SSSD only, using public key authentication.

### ProFTPD Version and Configuration

Note: /opt/proftpd is a symlink to /opt/proftpd-1.3.6c

```
/opt/proftpd-1.3.6c/sbin/proftpd -V
Compile-time Settings:
  Version: 1.3.6c (git)
  Platform: LINUX [Linux 3.10.0-1062.9.1.el7.x86_64 x86_64]
  Built: Mon Jan 13 2020 21:03:52 GMT
  Built With:
    configure  '--prefix=/opt/proftpd-1.3.6c' '--enable-ctrls' '--with-modules=mod_sftp:mod_tls:mod_ifsession:mod_ban:mod_auth_pam:mod_vroot'

  CFLAGS: -g2 -O2 -Wall -fno-omit-frame-pointer
  LDFLAGS: -L$(top_srcdir)/lib  -rdynamic 
  LIBS:  -lssl -lcrypto -L$(top_srcdir)/lib/libcap -lcap  -lpam  -lssl  -lcrypto -lsupp -lcrypt -ldl 

  Files:
    Configuration File:
      /opt/proftpd-1.3.6c/etc/proftpd.conf
    Pid File:
      /opt/proftpd-1.3.6c/var/proftpd.pid
    Scoreboard File:
      /opt/proftpd-1.3.6c/var/proftpd.scoreboard

  Info:
    + Max supported UID: 4294967295
    + Max supported GID: 4294967295

  Features:
    - Autoshadow support
    + Controls support
    - curses support
    - Developer support
    - DSO support
    + IPv6 support
    + Largefile support
    - Lastlog support
    - Memcache support
    - ncurses support
    - NLS support
    - Redis support
    - Sodium support
    + OpenSSL support (FIPS enabled)
    - PCRE support
    - POSIX ACL support
    + Shadow file support
    + Sendfile support
    + Trace support
    + xattr support

  Tunable Options:
    PR_TUNABLE_BUFFER_SIZE = 1024
    PR_TUNABLE_DEFAULT_RCVBUFSZ = 8192
    PR_TUNABLE_DEFAULT_SNDBUFSZ = 8192
    PR_TUNABLE_ENV_MAX = 2048
    PR_TUNABLE_GLOBBING_MAX_MATCHES = 100000
    PR_TUNABLE_GLOBBING_MAX_RECURSION = 8
    PR_TUNABLE_HASH_TABLE_SIZE = 40
    PR_TUNABLE_LOGIN_MAX = 256
    PR_TUNABLE_NEW_POOL_SIZE = 512
    PR_TUNABLE_PATH_MAX = 4096
    PR_TUNABLE_SCOREBOARD_BUFFER_SIZE = 80
    PR_TUNABLE_SCOREBOARD_SCRUB_TIMER = 30
    PR_TUNABLE_SELECT_TIMEOUT = 30
    PR_TUNABLE_TIMEOUTIDENT = 10
    PR_TUNABLE_TIMEOUTIDLE = 600
    PR_TUNABLE_TIMEOUTLINGER = 10
    PR_TUNABLE_TIMEOUTLOGIN = 300
    PR_TUNABLE_TIMEOUTNOXFER = 300
    PR_TUNABLE_TIMEOUTSTALLED = 3600
    PR_TUNABLE_XFER_SCOREBOARD_UPDATES = 10
```

OpenSSL is from openssl-1.0.2k-19.el7, openssl-devel-1.0.2k-19.el7:

```
# openssl version -a
OpenSSL 1.0.2k-fips  26 Jan 2017
built on: reproducible build, date unspecified
platform: linux-x86_64
options:  bn(64,64) md2(int) rc4(16x,int) des(idx,cisc,16,int) idea(int) blowfish(idx) 
compiler: gcc -I. -I.. -I../include  -fPIC -DOPENSSL_PIC -DZLIB -DOPENSSL_THREADS -D_REENTRANT -DDSO_DLFCN -DHAVE_DLFCN_H -DKRB5_MIT -m64 -DL_ENDIAN -Wall -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches   -m64 -mtune=generic -Wa,--noexecstack -DPURIFY -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_BN_ASM_GF2m -DRC4_ASM -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DMD5_ASM -DAES_ASM -DVPAES_ASM -DBSAES_ASM -DWHIRLPOOL_ASM -DGHASH_ASM -DECP_NISTZ256_ASM
OPENSSLDIR: "/etc/pki/tls"
engines:  dynamic 
```

Sanitized config and logs attached.  Both RSA and DSA keys are there (so one fails) from testing.

Thank you for taking a look at this!

[proftpdnd20.sanitized.log](https://github.com/proftpd/proftpd/files/4135891/proftpdnd20.sanitized.log)
[proftpext.sanitized.log](https://github.com/proftpd/proftpd/files/4135892/proftpext.sanitized.log)
[sftptrace.sanitized.log](https://github.com/proftpd/proftpd/files/4135893/sftptrace.sanitized.log)
[pam.d.configs.sanitized.txt](https://github.com/proftpd/proftpd/files/4135912/pam.d.configs.sanitized.txt)
[proftpd.conf.sanitized.txt](https://github.com/proftpd/proftpd/files/4135910/proftpd.conf.sanitized.txt)

**Comment by Castaglia** [comment]
What kind of public key is being used?  I.e. is it a DSA key, an RSA key, an EC key, or...?  I ask because if it's a DSA key in particular, it may be fixed by [this commit](https://github.com/proftpd/proftpd/commit/8edd398648013ac401e9b9fdcdfdd4f8a99e3b27).

**Comment by jpfinger** [comment]
I am getting this with both RSA and DSA keys.  That was one of the first things I messed with in case it was algo related.  The authorized_keys/%u file has both (converted) keys in it.

The source I'm compiling from already has that commit.

**Comment by Castaglia** [comment]
In the provided ProFTPD debug logging (thanks!), I see this:
```
2020-01-30 18:49:38,631 myhostname proftpd[15300] X.X.X.Y (X.X.X.C[X.X.X.C]): USER PubKeyUser (Login failed): User not authorized for login
```
This suggests (based on an earlier log message mentioning `getspnam(3)`) that the account (at least for this user) is locked for some reason.

**Comment by jpfinger** [comment]
I can login with the same user if I use password authentication.  Also, if I revert to our previous Proftpd 1.3.5 the publickey auth works.  In that case all else is the same (same FTP server, same SSSD, etc).

**Comment by Castaglia** [comment]
The check for expired accounts, via shadow password information, was added to 1.3.6 in [this commit](https://github.com/proftpd/proftpd/commit/757b9633191eafa32a86ab8ec032e743d0227093), for  [Bug #4308](http://bugs.proftpd.org/show_bug.cgi?id=4308).  That's why this wasn't happening with ProFTPD 1.3.5.  I'm now researching why you aren't encountering this behavior for password authentication.

**Comment by Castaglia** [comment]
OK, I see why this is login failure is happening for SFTP publickey authentication, but *not* SFTP password authentication.

In `mod_auth_unix.c#pw_auth` (used for password authentication), we see that a lack of shadow password information is handled by returning `DECLINED` [here](https://github.com/proftpd/proftpd/blob/master/modules/mod_auth_unix.c#L728):
```
  cleartxt_passwd = get_pwd_info(cmd->tmp_pool, name, &lstchg, NULL, &max,
    NULL, &inact, &expire);
  if (cleartxt_passwd == NULL) {
    return PR_DECLINED(cmd);
  }
```
And a `DECLINED` here does not cause issues; it merely indicates the handler declines to handle things.

But in `mod_auth_unix.c#pw_authz` (used for publickey authentication), the lack of shadow password information is handled by returning `ERROR` [here](https://github.com/proftpd/proftpd/blob/master/modules/mod_auth_unix.c#L772):
```
cleartxt_passwd = get_pwd_info(cmd->tmp_pool, user, &lstchg, NULL, &max,
    NULL, &inact, &expire);
  if (cleartxt_passwd == NULL) {
    pr_log_auth(LOG_WARNING, "no password information found for user '%.100s'",
      user);
    return PR_ERROR_INT(cmd, PR_AUTH_NOPWD);
  }
```
I'll put a PR that makes `pw_authz` return `DECLINED` for this situation as well, to see if that helps address the issue.

**Comment by jpfinger** [comment]
Thanks for the quick patch.  After the update the publickey auth works, in addition to normal username+password on the account.

**Comment by Castaglia** [comment]
Excellent, thanks for confirming!  PR merged to master, and backported to the 1.3.6 branch.
